Normalizing Each Objective Class's motionCost()

akash-venkateshwaran opened the pull request:
<!-- Remember to add the relevant reviewers, assignees, and labels, and create this PR as a draft if it is a work in progress. -->

### Description
<!-- Replace _issue_ with the issue number that this PR resolves, or delete the line and add this PR to the Software project if there is no related issue. -->
- Resolves UBCSailbot/sailbot_workspace#290 
<!-- Describe what was done in this PR if there is no related issue, or the related issue's description is not sufficient. -->
- Added two functions in the parent class: (i) Creates random sample states; (ii) Compares every pair in that sample and picks the maximum motionCost()
- Inherited these functions in each Objective class to find the max_motionCost for each objective function
- Modified the motionCost() function of each Objective class to normalise its cost using max_motionCost

### Verification
<!-- List the steps that were taken to verify that the changes introduced by this PR function as desired and without side effects. -->
- Verified using a simple setup example problem
- Haven't included unit tests

### Resources
<!-- Link to any resources that are relevant to this PR. -->
-

akash-venkateshwaran committed:
Added normalising funcs to Distance Obj

akash-venkateshwaran committed:
Corrected the code for SE2StateSpace()

akash-venkateshwaran committed:
Shifted the def of unc to parent class and inherited in all the objectives

patrick-5546 committed:
Add missing type hint

patrick-5546 committed:
Remove _ prefix

patrick-5546 committed:
Refactor normalization

patrick-5546 committed:
sample and find max cost in base objective

patrick-5546 committed:
Code style and docstrings

patrick-5546 committed:
Add tests

patrick-5546 commented on local_pathfinding/objectives.py:
```diff
@@ -64,14 +104,21 @@ class DistanceObjective(Objective):
+        space_information,
+        method: DistanceMethod,
+        reference: cs.LatLon = cs.LatLon(0, 0),
+        num_samples: int = 100,
```
If you don't think `num_samples` won't change dynamically, we can remove this argument and just hardcode when calling the super's init function

```python
super().__init__(space_information, num_samples=100)
```

akash-venkateshwaran replied:
Yup, num_sample won't change dynamically but I think it will change for each objective function. A more complicated objective function would require more samples to find a close estimate of the max_motionCost. Sure, we can hardcode it.

patrick-5546 reviewed (commented):
I added some tests in `test_distance_objective` that we should use for the other objectives:
- `sample_states()`: commented out checking that all states are valid, which I think is what we want right?
- `find_maximum_motion_cost()`
    - Implicitly test by comparing `max_motion_cost` to a hardcoded value
    - Tried to test `find_maximum_motion_cost()` directly but was not able to get the type of `states` correct so I commented it out for now
        - Update: fixed and uncommented

If you want to merge this PR today and aren't able to finish the tests, I'm ok with merging this PR and working on the tests in another PR. Just leave some `# TODO: <message>` comments

patrick-5546 committed:
Fix explicit test

patrick-5546 committed:
Cleanup

patrick-5546 committed:
Formatting

akash-venkateshwaran commented:
> I added some tests in `test_distance_objective` that we should use for the other objectives:

> 
> * `sample_states()`: commented out checking that all states are valid, which I think is what we want right?
> * `find_maximum_motion_cost()`
>   
>   * Implicitly test by comparing `max_motion_cost` to a hardcoded value
>   * Tried to test `find_maximum_motion_cost()` directly but was not able to get the type of `states` correct so I commented it out for now
>     
>     * Update: fixed and uncommented
> If you want to merge this PR today and aren't able to finish the tests, I'm ok with merging this PR and working on the tests in another PR. Just leave some `# TODO: <message>` comments

Thanks for adding the unit tests.
Yes we want to make sure all the sampled states are valid which means they are not in the collision area
Great! So now I have to repeat these tests for other objective functions right?

patrick-5546 commented:
> Great! So now I have to repeat these tests for other objective functions right?

Yes and ensure that they are the sampled states are valid. The validity checker currently is just a placeholder but it will be good to set this up now anyways.

akash-venkateshwaran committed:
Added unit tests for all three objs

patrick-5546 committed:
Merge remote-tracking branch 'origin' into user/akash-venkateshwaran/33-normalising_obj_funcs

patrick-5546 committed:
Hardcode num_samples in each objective function

patrick-5546 committed:
Update docstrings

patrick-5546 committed:
Reorder tests

patrick-5546 commented on test/test_objectives.py:
```diff
@@ -20,20 +22,54 @@
+    # test find_maximum_motion_cost()
+    # implicitly
+    assert distance_objective.max_motion_cost == pytest.approx(max_motion_cost, rel=1e0)
+    # explicitly for the easiest method to set up
+    # don't need to test for all methods since they each have their own tests
+    if method == objectives.DistanceMethod.OMPL_PATH_LENGTH:
+        states = []
+        space = PATH._simple_setup.getStateSpace()
+        for xy in [(-0.5, 0.4), (0.1, 0.2), (0.3, -0.6)]:
+            state = ob.State(space)
+            state().setXY(*xy)
+            states.append(state)
+        max_cost = max(
+            distance_objective.ompl_path_objective.motionCost(s1(), s2()).value()
+            for s1, s2 in itertools.combinations(iterable=states, r=2)
+        )
+        assert distance_objective.find_maximum_motion_cost([s() for s in states]) == pytest.approx(
+            max_cost / distance_objective.max_motion_cost
+        )
```
What do you think about adding an explicit test for the other objectives? Only need to test one method per objective

akash-venkateshwaran replied:
Sure but I didn't understand why we are doing this. I think we are checking if find_maximum_motion_cost() gives the same output as max() for three samples. But find_maximum_motion_cost() itself computes it using max() function right?

patrick-5546 replied:
Yea this seems unnecessary, removed

patrick-5546 commented on test/test_objectives.py:
```diff
@@ -20,20 +22,54 @@
+    num_samples = 3
+    sampled_states = distance_objective.sample_states(num_samples)
+    assert len(sampled_states) == num_samples
+    # for state in sampled_states:
+    #     assert ompl_path.is_state_valid(state)
```
We discussed this a while ago, but I think sample_states should return valid states. Then you can uncomment these 2 lines from all the new tests

akash-venkateshwaran replied:
Sure, but it will fail cause the validity check is not implemented in the sampler. I think Sean has to first define the target boats and islands for us to implement this. Let me know if I am wrong. It can easily be added I just have to add that check before appending to the list within sampler function.

patrick-5546 replied:
You can use the function in this commented out code, `is_state_valid()`

patrick-5546 replied:
Yea no need to do this as we discussed, the cost doesn't take into account the state validity anyways

patrick-5546 reviewed (commented):
It seems like sometimes the normalized cost is greater than 1: https://github.com/UBCSailbot/local_pathfinding/actions/runs/7688802043/job/20950435771

patrick-5546 committed:
Remove redundant check

patrick-5546 committed:
Fix lint error

akash-venkateshwaran committed:
Added cap func for motionCost

patrick-5546 commented on local_pathfinding/objectives.py:
```diff
@@ -40,19 +41,64 @@ class Objective(ob.StateCostIntegralObjective):
+
+        return sampled_states
+
+    def capping_motionCost(self, value):
+        if value > 1:
+            value == 1
+        return value
```
Let's make a function that just returns the normalized cost with capping. So normalize the value with `self.max_motion_cost` before capping and returning

patrick-5546 replied:
Also don't forget the docstring

patrick-5546 replied:
Akash has done this

patrick-5546 commented on local_pathfinding/objectives.py:
```diff
@@ -40,19 +41,64 @@ class Objective(ob.StateCostIntegralObjective):
         self.space_information = space_information
 
+        states = self.sample_states(num_samples)
+        self.max_motion_cost = 1.0
+        self.max_motion_cost = self.find_maximum_motion_cost(states)
```
I forgot, is there a reason why you initialize it to 1 first? Would it be possible to do this inside `find_maximum_motion_cost()`?

akash-venkateshwaran replied:
Yes, we divide the motionCost() by max_motion_cost. So initially when we don't know the max_motion_cost we divide it as 1. Then we sample states and find the max_motion_cost. Finally normalise using the new max_motion_cost we obtain from the sample

patrick-5546 replied:
Added a comment to explain this

akash-venkateshwaran committed:
Replaced capping with a single function named normalization

patrick-5546 committed:
Simplify normalization logic

patrick-5546 committed:
Add explanatory comment

patrick-5546 commented:
As discussed, just need to resolve merge conflicts:
- `git merge origin main`
- Add speed objective
- Normalize speed objective
- Update weights

akash-venkateshwaran committed:
wokring on SpeedObj

akash-venkateshwaran committed:
wokring on SpeedObj

patrick-5546 committed:
Fix test and cleanup

jamenkaye committed:
dynamically compute heading s1 to s2

patrick-5546 committed:
Merge branch 'main' into user/akash-venkateshwaran/33-normalising_obj_funcs
